File: src/main/java/ai/ancf/lmos/operator/server/routing/impl/DefaultCustomResourcesService.java
/*
 * SPDX-FileCopyrightText: 2024 Deutsche Telekom AG
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package ai.ancf.lmos.operator.server.routing.impl;

import io.fabric8.kubernetes.client.KubernetesClient;
import ai.ancf.lmos.operator.resources.channel.ChannelResource;
import ai.ancf.lmos.operator.resources.routing.ChannelRoutingResource;
import ai.ancf.lmos.operator.server.routing.CustomResourcesService;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@Component
public class DefaultCustomResourcesService implements CustomResourcesService {

    private final KubernetesClient client;

    public DefaultCustomResourcesService(KubernetesClient client) {
        this.client = client;
    }
    @Override
    public Optional<ChannelRoutingResource> getRouting(String tenant, String channel, String subset) {
        Map<String, String> labelSelectors = Map.of(
                "tenant", tenant,
                "channel", channel,
                "subset", subset
        );

        var channelRoutingResources = client.resources(ChannelRoutingResource.class).withLabels(labelSelectors).list();

        if(channelRoutingResources.getItems().isEmpty()){
           return Optional.empty();
        }else {
            ChannelRoutingResource channelRoutingResource = channelRoutingResources.getItems().get(0);
            channelRoutingResource.getMetadata().setOwnerReferences(null);
            channelRoutingResource.getMetadata().setManagedFields(null);
            return Optional.of(channelRoutingResource);
        }
    }

    @Override
    public List<ChannelResource> getChannels(String tenant, String subset) {
        Map<String, String> labelSelectors = Map.of(
                "tenant", tenant,
                "subset", subset
        );

        var channelResources = client.resources(ChannelResource.class).withLabels(labelSelectors).list();
        channelResources.getItems().forEach(channelResource -> channelResource.getMetadata().setManagedFields(null));
        return channelResources.getItems();
    }

    @Override
    public Optional<ChannelResource> getChannel(String tenant, String channel, String subset) {
        Map<String, String> labelSelectors = Map.of(
                "tenant", tenant,
                "channel", channel,
                "subset", subset
        );

        var channelResources = client.resources(ChannelResource.class).withLabels(labelSelectors).list();

        if(channelResources.getItems().isEmpty()){
            return Optional.empty();
        }else {
            ChannelResource channelResource = channelResources.getItems().get(0);
            channelResource.getMetadata().setManagedFields(null);
            return Optional.of(channelResource);
        }
    }
}

File: src/main/kotlin/ai/ancf/lmos/operator/server/routing/impl/DefaultCustomResourcesService.kt
/*
 * SPDX-FileCopyrightText: 2024 Deutsche Telekom AG
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.eclipse.lmos.operator.server.routing.impl

import io.fabric8.kubernetes.client.KubernetesClient
import org.eclipse.lmos.operator.resources.ChannelResource
import org.eclipse.lmos.operator.resources.ChannelRoutingResource
import org.eclipse.lmos.operator.server.routing.CustomResourcesService
import org.springframework.stereotype.Component
import java.util.function.Consumer

@Component
class DefaultCustomResourcesService(private val client: KubernetesClient) : CustomResourcesService {
    override fun getRouting(
        tenant: String,
        channel: String,
        subset: String,
        namespace: String,
    ): ChannelRoutingResource? {
        val labelSelectors =
            mapOf(
                "tenant" to tenant,
                "channel" to channel,
                "subset" to subset,
            )

        val channelRoutingResources =
            client.resources(
                ChannelRoutingResource::class.java,
            ).inNamespace(namespace).withLabels(labelSelectors).list()

        if (channelRoutingResources.items.isEmpty()) {
            return null
        } else {
            val channelRoutingResource = channelRoutingResources.items[0]
            channelRoutingResource.metadata.ownerReferences = null
            channelRoutingResource.metadata.managedFields = null
            return channelRoutingResource
        }
    }

    override fun getChannels(
        tenant: String,
        subset: String,
        namespace: String,
    ): List<ChannelResource> {
        val labelSelectors =
            mapOf(
                "tenant" to tenant,
                "subset" to subset,
            )

        val channelResources =
            client.resources(
                ChannelResource::class.java,
            ).inNamespace(namespace).withLabels(labelSelectors).list()
        channelResources.items.forEach(
            Consumer { channelResource: ChannelResource ->
                channelResource.metadata.managedFields = null
            },
        )
        return channelResources.items
    }

    override fun getChannel(
        tenant: String,
        channel: String,
        subset: String,
        namespace: String,
    ): ChannelResource? {
        val labelSelectors =
            mapOf(
                "tenant" to tenant,
                "channel" to channel,
                "subset" to subset,
            )

        val channelResources =
            client.resources(
                ChannelResource::class.java,
            ).inNamespace(namespace).withLabels(labelSelectors).list()

        if (channelResources.items.isEmpty()) {
            return null
        } else {
            val channelResource = channelResources.items[0]
            channelResource.metadata.managedFields = null
            return channelResource
        }
    }
}
